### PR TITLE
Document new style function declarations

### LINKED ISSUE #722: Arrow-style function declarations aren't covered in the Mir style guide
For example:

```c++
template<typename... Extensions>
auto with_extension(WaylandExtensions const& wayland_extensions,
                   std::string const& name, std::function<std::shared_ptr<void>(wl_display*)> builder,
                   Extensions... extensions) -> WaylandExtensions
{
    return with_extension(with_extension(wayland_extensions, name, builder), extensions...);
}
```

[The style guide section on function declarations](https://mir-server.io/doc/cppguide/#Function_Declarations_and_Definitions) does not cover how to properly format this code.

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Document new style function declarations. (Fixes #722)

### CODE REVIEW

File: guides/cppguide.xml [hunk 1]
```diff
@@ -3054,68 +3054,69 @@ Thomas Voss
...
         {
             ...
         }
 
         // This function signature requires multiple lines, but
         // the const keyword is on the line with the last parameter.
-        ReturnType really_long_function_name(Type par1,
-                                             Type par2) const
+        void really_long_function_name(Type par1,
+                                       Type par2) const
```
Comment (RAOF) [inline guides/cppguide.xml]: Could you kindly fix the indentation of the function arguments while you're here?

File: guides/cppguide.xml [hunk 2]
```diff
@@ -3054,68 +3054,66 @@ Thomas Voss
...
-             </li>
+       </li>
         <li> All parameters should be named, with identical names in the
              declaration and implementation. (Except where unused parameters
              are suppressed in the implementation.)
-             </li>
+       </li>
         <li> All parameters should be in a single line if possible, otherwise:
             <ul>
-                <li> Wrap groups of parameters into the next line, and align the
```
Comment (RAOF) [inline guides/cppguide.xml]: We're not doing grouping at all anymore?

Comment (AlanGriffiths) [reply guides/cppguide.xml]: We're aligning by an indent, not with the first parameter

Comment (RAOF) [reply guides/cppguide.xml]: I still think this removes the “you may group related parameters” bit?

**Pull request merged**